services/workflows-service/src/workflow/workflow.service.ts

```diff
@@ -648,38 +611,124 @@ export class WorkflowService {
-            status,
-          },
-        };
+    const documentWithDecision = {
+      ...document,
+      id: document.id,
+      decision: {
+        ...newDecision,
+        status,
       },
+    };
+    const checkRequiredFields = status === 'approved' ? true : false;
+    const updatedWorkflow = await this.updateDocumentById(
+      { workflowId, documentId, checkRequiredFields },
+      documentWithDecision as unknown as DefaultContextSchema['documents'][number],
+    );
+
+    return updatedWorkflow;
+  }
+  async updateDocumentById(
+    {
+      workflowId,
+      documentId,
+      checkRequiredFields = true,
+    }: {
+      workflowId: string;
+      documentId: string;
+      checkRequiredFields?: boolean;
+    },
+    data: DefaultContextSchema['documents'][number] & { propertiesSchema?: object },
+  ) {
+    const runtimeData = await this.workflowRuntimeDataRepository.findById(workflowId);
+    const workflowDef = await this.workflowDefinitionRepository.findById(
+      runtimeData.workflowDefinitionId,
     );
+    const document = {
+      ...data,
+      id: documentId,
+    };
+
+    const newDocument = addPropertiesSchemaToDocument(document);
+    const propertiesSchema = newDocument?.propertiesSchema ?? {};
+    if (Object.keys(propertiesSchema)?.length) {
+      let propertiesSchemaForValidation = propertiesSchema;
+      if (!checkRequiredFields) {
+        const { required: _required, ..._propertiesSchemaForValidation } = propertiesSchema;
+        propertiesSchemaForValidation = _propertiesSchemaForValidation;
+      }
+      const validatePropertiesSchema = ajv.compile(propertiesSchemaForValidation);
+
+      const isValidPropertiesSchema = validatePropertiesSchema(newDocument?.properties);
+
+      if (!isValidPropertiesSchema) {
+        throw new BadRequestException(
+          validatePropertiesSchema.errors?.map(({ instancePath, message, ...rest }) => ({
+            ...rest,
+            message: `${instancePath} ${message}`,
+            instancePath,
+          })),
+        );
+      }
+    }
+
     const updatedWorkflow = await this.updateContextById(workflowId, {
-      documents: documentsWithDecision,
+      documents: [newDocument],
     });
-
     const correlationId = await this.getCorrelationIdFromWorkflow(updatedWorkflow);
 
+    if (
+      ['active'].includes(updatedWorkflow.status) &&
+      workflowDef.config?.completedWhenTasksResolved
+    ) {
+      const allDocumentsResolved =
+        updatedWorkflow.context?.documents?.length &&
+        updatedWorkflow.context?.documents?.every(
+          (document: DefaultContextSchema['documents'][number]) => {
+            return ['approved', 'rejected', 'revision'].includes(
+              document?.decision?.status as string,
+            );
+          },
+        );
+
+      updatedWorkflow.status = allDocumentsResolved ? 'completed' : updatedWorkflow.status;
+      await this.workflowRuntimeDataRepository.updateById(workflowId, {
```

Comment:
Using the return value of `updateById` over mutating `updatedWorkflow` seems safer for the long term, over time `updateById` may change data differently.